Lib/test/test_ast/test_ast.py

```diff
@@ -3175,23 +3179,168 @@ def test_subinterpreter(self):
         self.assertEqual(res, 0)
 
 
-class ASTMainTests(unittest.TestCase):
-    # Tests `ast.main()` function.
+class CommandLineTests(unittest.TestCase):
+    def setUp(self):
+        self.filename = tempfile.mktemp()
+        self.addCleanup(os_helper.unlink, self.filename)
 
-    def test_cli_file_input(self):
-        code = "print(1, 2, 3)"
-        expected = ast.dump(ast.parse(code), indent=3)
-
-        with os_helper.temp_dir() as tmp_dir:
-            filename = os.path.join(tmp_dir, "test_module.py")
-            with open(filename, 'w', encoding='utf-8') as f:
-                f.write(code)
-            res, _ = script_helper.run_python_until_end("-m", "ast", filename)
+    @staticmethod
+    def text_normalize(string):
+        return textwrap.dedent(string).strip()
+
+    def set_source(self, content):
+        Path(self.filename).write_text(self.text_normalize(content))
+
+    def invoke_ast(self, *flags):
+        stderr = StringIO()
+        stdout = StringIO()
+        with (
+            contextlib.redirect_stdout(stdout),
+            contextlib.redirect_stderr(stderr),
+        ):
+            ast.main(args=[*flags, self.filename])
+        self.assertEqual(stderr.getvalue(), '')
+        return self.text_normalize(stdout.getvalue())
+
+    def check_output(self, source, expect, *flags):
+        with self.subTest(source=source, flags=flags):
+            self.set_source(source)
+            res = self.invoke_ast(*flags)
+            expect = self.text_normalize(expect)
+            self.assertEqual(res, expect)
+
+    def test_invocation(self):
+        # test various combinations of parameters
+        base_flags = (
+            ('-m=exec', '--mode=exec'),
+            ('--no-type-comments', '--no-type-comments'),
+            ('-a', '--include-attributes'),
+            ('-i=4', '--indent=4'),
+        )
+        self.set_source('''
+            print(1, 2, 3)
+            def f(x: int) -> int:
+                x -= 1
+                return x
+        ''')
 
-        self.assertEqual(res.err, b"")
-        self.assertEqual(expected.splitlines(),
-                         res.out.decode("utf8").splitlines())
-        self.assertEqual(res.rc, 0)
+        for r in range(1, len(base_flags) + 1):
+            for choices in itertools.combinations(base_flags, r=r):
+                for args in itertools.product(*choices):
+                    with self.subTest(flags=args[1:]):
+                        _ = self.invoke_ast(*args)
+
+    def test_unknown_flag(self):
+        with self.assertRaises(SystemExit):
+            output = self.invoke_ast('--unknown')
+            self.assertStartsWith(output, 'usage: ')
```

Comment:
this test seems incorrect. do we really have to assert something in `with self.assertRaises(SystemExit):`?

Reply:
without `assertRaises` tests not passing and we getting SystemExit: 0 and SystemExit: 2 
(and other CLI test also use this pattern)

Reply:
I am ok with `assertRaises`, `self.assertStartsWith(output, 'usage: ')` seems strange.

Reply:
I guess it could be useful if you want to assert that the help text is printed on error? Though I'd dedent the assert so that it is outside the with block.

Reply:
unfortunately we don't have access to the `output` object outside of this block, but i'm already rewrite this test.
I am combining tests with help message because the general logic of catching messages